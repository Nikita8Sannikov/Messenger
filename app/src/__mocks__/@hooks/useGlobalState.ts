export const mockUseAppSelector = jest.fn();
export const mockUseAppDispatch = jest.fn();