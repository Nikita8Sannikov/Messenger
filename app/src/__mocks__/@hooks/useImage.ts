const useImage = (src: string | null): string => {
	return src || "/assets/images/noPhoto.jpg";
};

export default useImage; 