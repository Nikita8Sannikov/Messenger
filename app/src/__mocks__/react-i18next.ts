export const useTranslation = () => ({
    t: (key: string) => {
        const dictionary: Record<string, string> = {
            "menu.profile": " My profile",
            "menu.messanger": "Messenger",
            "menu.friends": "Friends",
            "menu.to-developers": "To developers",
            "header.settings": "Settings",
            "header.help": "Help",
            "header.logout": "Log Out"
        };
        return dictionary[key];
    },
    i18n: {
        changeLanguage: jest.fn(),
        use: jest.fn().mockReturnThis(), // Мок для вызова use()
        init: jest.fn(), // Мок для вызова init()
    },
})

export const initReactI18next = {
    type: "3rdParty",
    init: jest.fn(),
}