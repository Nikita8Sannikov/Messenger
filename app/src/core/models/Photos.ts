import { type IPhoto } from "@custom-types/models.types";

// Контракт модели "Фотографии"
export interface Photos {
	photos: IPhoto[];
	count: number;

	getAllPhotos: (resolve?: () => void) => void;
	onClickPhoto: (photoId: string) => void;
	addPhotoFromAvatar: (photo: IPhoto) => void;
	addPhotos: (data: FormData) => void;
	deletePhoto: (data: { photoId: string; imageUrl: string; isAvatar: boolean; }, fromProfile?: boolean) => void;
};