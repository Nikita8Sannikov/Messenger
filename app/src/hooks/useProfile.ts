import useMainClient from "@hooks/useMainClient";

// Возврат сущности "Профиль" конкретного пользователя по переданному идентификатору
export default function useProfile(userId?: string) {
	const mainClient = useMainClient();
	return mainClient.getProfile(userId);
}
