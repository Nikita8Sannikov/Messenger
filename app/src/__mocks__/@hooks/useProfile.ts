import { type IPhoto } from "@custom-types/models.types";

const mockNotificationSettingsService = {
	soundEnabled: false,
	messageSound: false,
	friendRequestSound: false,
	changeSoundNotifications: jest.fn(),
};

const mockPhotosService = {
	photos: [] as IPhoto[],
	count: 0,
	getAllPhotos: jest.fn(),
	onClickPhoto: jest.fn(),
	addPhotoFromAvatar: jest.fn(),
	addPhotos: jest.fn(),
	deletePhoto: jest.fn(),
};

const mockUserDetailsService = {
	updateDetails: jest.fn(),
	syncUserDetails: jest.fn(),
	getFriendsText: jest.fn(),
	getSubscribersText: jest.fn(),
	getPhotosText: jest.fn(),
	getAudiosText: jest.fn(),
	getVideosText: jest.fn(),
};

const mockUserService = {
	photosService: mockPhotosService,
	detailsService: mockUserDetailsService,
	settingsService: mockNotificationSettingsService,
	id: "",
	firstName: "",
	thirdName: "",
	phone: "",
	email: "",
	fullName: "",
	avatarUrl: "",
	avatarCreateDate: "",
	updateUser: jest.fn(),
	syncInfo: jest.fn(),
	changeAvatar: jest.fn(),
};

const mockProfileService = {
	isMe: false,
	userService: mockUserService,
	photosService: mockPhotosService,
	deleteAccount: jest.fn(),
	onSetAvatar: jest.fn(),
	onDeleteAvatar: jest.fn(),
	onClickAvatar: jest.fn(),
	getFriends: jest.fn(),
	editInfo: jest.fn(),
};

const useProfile = jest.fn().mockReturnValue(mockProfileService);

export default useProfile;
export { 
	mockProfileService, 
	mockUserService, 
	mockPhotosService, 
	mockUserDetailsService, 
	mockNotificationSettingsService, 
};