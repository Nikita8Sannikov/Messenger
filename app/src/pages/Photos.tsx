import { type ChangeEvent, useCallback, useRef, useState } from "react";
import { type ListOnScrollProps } from "react-window";

import BoxComponent from "@components/ui/box";
import InputImageComponent from "@components/ui/input-image";
import PaperComponent from "@components/ui/paper";
import PhotoComponent from "@components/ui/photo";
import UpperButton from "@components/ui/upper-button";
import useUser from "@hooks/useUser";
import VirtualList, { type VirtualListHandle } from "@modules/virtual-list/list";
import i18next from "@service/i18n";
import usePhotosStore from "@store/photos";
import { type IPhoto } from "@custom-types/models.types";

import "@styles/pages/photos.scss";

const SCROLL_TRESHHOLD = 600;
const ITEM_HEIGHT = 250;    // Такая же высота необходима быть указана в scss файле для высоты .photo
const COLS = 3;
const GAP = 8;

// Страница "Фотографии"
export default function Photos() {
	const [ showUpButton, setShowUpButton ] = useState(false);

	const containerRef = useRef<HTMLDivElement | null>(null);
	const virtualRef = useRef<VirtualListHandle | null>(null);

	const photos = usePhotosStore(state => state.photos);
	const isLoading = usePhotosStore(state => state.isPhotosLoading);
	const hasMore = usePhotosStore(state => state.hasMore);
	const isAddPhotosLoading = usePhotosStore(state => state.isAddPhotosLoading);

	const photosService = useUser().photosService;

	/**
     * Обработчик, вызываемый при изменении видимых элементов списка.
     * Отдаём обещание, чтобы InfiniteLoader знал, когда закончился запрос.
     * Также, эта функция выполняется при первоначальной подгрузке данных, когда экран может быть очень высоким,
     * а количество подгружаемых элементов (PHOTOS_LIMIT) настроено на маленькое число.
     */
	const loadMoreItems = useCallback(() => {
		if (!hasMore || isLoading) {
			return Promise.resolve();
		}

		// Возвращаем промис, который завершится в successCb
		return new Promise<void>(() => {
			photosService.getAllPhotos();
		});
	}, [ hasMore, isLoading ]);

	// Обработка скролла
	const handleScroll = useCallback(({ scrollOffset }: ListOnScrollProps) => {
		// Показываем кнопку "Вверх" после 200px прокрутки
		setShowUpButton(scrollOffset > SCROLL_TRESHHOLD);
	}, []);

	// Скролл до самого верха
	const scrollToTop = () => {
		virtualRef.current?.scrollTop();
	};

	// Явно задаем, что должен отрисовать на каждой строке виртуальный список
	const renderCb = useCallback(({ item, itemWidth, itemHeight }: { item: IPhoto; itemWidth: number; itemHeight: number; }) => {
		return <div key={item.id} className="photos__container__item" style={{ width: itemWidth, height: itemHeight }}>
			<PhotoComponent
				src={`${item.path}?w=${itemWidth * 2}`}
				alt={item.id}
				isLazy
				clickHandler={() => photosService.onClickPhoto(item.id)}
				deleteHandler={() => photosService.deletePhoto({
					photoId: item.id,
					imageUrl: item.path,
					isAvatar: false,
				})}
			/>
		</div>;
	}, []);

	// Добавление новых фотографий
	const addPhotosHandler = (e: ChangeEvent<HTMLInputElement>) => {
		const fileList = e.target.files;

		if (fileList && fileList.length) {
			const formData = new FormData();
			const files = Array.from(fileList);

			files.forEach((file) => {
				formData.append("photo", file);
			});

			photosService.addPhotos(formData);
		}
	};

	return <PaperComponent className="photos paper-block">
		<BoxComponent ref={containerRef} className="photos__container">
			<InputImageComponent
				id="photos-add-new-photos"
				text={i18next.t("photos-module.add_more")}
				loading={isAddPhotosLoading}
				multiple
				onChange={addPhotosHandler}
			/>

			<VirtualList<IPhoto>
				ref={containerRef}
				virtualRef={virtualRef}
				items={photos}
				options={{ cols: COLS, gap: GAP, itemHeight: ITEM_HEIGHT }}
				hasMore={hasMore}
				emptyText={i18next.t("photos-module.empty")}
				isLoading={isLoading}
				loadMore={loadMoreItems}
				renderCb={renderCb}
				onScroll={handleScroll}
			/>

			{showUpButton ? <UpperButton onClick={scrollToTop} /> : null}
		</BoxComponent>
	</PaperComponent>;
};