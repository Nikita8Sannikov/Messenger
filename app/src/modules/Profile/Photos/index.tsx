import { useEffect,useState } from "react";
import { useTranslation } from "react-i18next";
import { useNavigate } from "react-router-dom";

import GridComponent from "@components/ui/grid";
import InputImageComponent from "@components/ui/input-image";
import LinkComponent from "@components/ui/link";
import NoDataComponent from "@components/ui/no-data";
import PaperComponent from "@components/ui/paper";
import PhotoComponent from "@components/ui/photo";
import SpinnerComponent from "@components/ui/spinner";
import { useAppSelector } from "@hooks/useGlobalState";
import useProfile from "@hooks/useProfile";
import { selectUserState } from "@store/user/slice";
import { Pages } from "@custom-types/enums";

import "./photos.scss";

export default function Photos() {
  const [loadingPhotos, setLoadingPhotos] = useState(false);
  const [loadingBtn, setLoadingBtn] = useState(false);

  const profile = useProfile();
  const { t } = useTranslation();
  const { photos } = useAppSelector(selectUserState);
  const navigate = useNavigate();

  // Получаем все фотографии пользователя
  useEffect(() => {
    profile.getAllPhotos(setLoadingPhotos);
  }, []);

  // Обновляем счетчик фотографий
  useEffect(() => {
    profile.updatePhotosCount(photos.length);
  }, [photos]);

  // Клик по одной из своих фотографий
  const onClickPhoto = (index: number) => {
    profile.onClickPhoto(photos, index);
  };

  // Добавление новых фотографий
  const addPhotosHandler = async (e: React.ChangeEvent<HTMLInputElement>) => {
    const fileList = e.target.files;

    if (fileList && fileList.length) {
      const formData = new FormData();
      const files = Array.from(fileList);

      files.forEach((file) => {
        formData.append("photo", file);
      });

      profile.addPhotos(formData, setLoadingBtn);
    }
  };

  // Удаление одной фотографии
  const deleteOnePhoto = (photoPath: string) => {
    profile.deletePhoto(
      { imageUrl: photoPath, isAvatar: false },
      photos,
      photoPath
    );
  };

  return (
    <GridComponent className="photo-container__grid">
      <PaperComponent className="photo-container paper-block">
        <div className="photo-container__title">
          <div className="block-title photo-container__title__text">
            {t("profile-module.my_photos")}
            <span className="counter">{photos.length}</span>
          </div>

          <LinkComponent
            variant="body2"
            onClick={() => navigate(Pages.photos)}
            underline="hover"
            className="photo-container__title__link"
          >
            {t("profile-module.watch_all")}
          </LinkComponent>
        </div>

        {
          loadingPhotos
            ? <SpinnerComponent />
            : photos && photos.length
              ? <div className="photo-container__photos">
                {photos.slice(0, 3).map((photo, index) => {
                  return <div key={photo.id} className="photo-container__photo-wrapper">
                    <PhotoComponent
                      key={photo.id}
                      src={photo.path}
                      alt={profile.user.fullName + " " + index}
                      clickHandler={() => onClickPhoto(index)}
                      deleteHandler={() => deleteOnePhoto(photo.path)}
                    />
                  </div>
                })}
              </div>
              : <NoDataComponent text={t("profile-module.no_photos")} />
        }
        
        <InputImageComponent
          id="add-new-photos"
          text={t("profile-module.add_more")}
          loading={loadingBtn}
          multiple
          onChange={addPhotosHandler}
        />
      </PaperComponent>
    </GridComponent>
  );
};