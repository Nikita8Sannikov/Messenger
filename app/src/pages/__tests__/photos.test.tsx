import { act, fireEvent, render } from "@testing-library/react";

import Photos from "@pages/Photos";
import type { IPhoto } from "@custom-types/models.types";
import { mockPhotosService } from "../../__mocks__/@hooks/useProfile";
import mockPhotosStore from "../../__mocks__/@store/photos";
import mockProfileStore from "../../__mocks__/@store/profile";
import mockUserStore from "../../__mocks__/@store/user";

import "../../__mocks__/react-i18next";

jest.mock("@hooks/usePrepareAnotherUser", () => ({
	__esModule: true,
	default: () => ({
		isLoadingAnotherUser: false,
		isExistProfile: true,
	}),
}));

describe("Photos Component", () => {
	const mockPhotos: IPhoto[] = [
		{
			id: "1",
			userId: "123",
			size: "1024",
			path: "/path/to/photo1.jpg",
			extension: "jpg",
			createdAt: "2024-01-01",
		},
		{
			id: "2",
			userId: "123",
			size: "1024",
			path: "/path/to/photo2.jpg",
			extension: "jpg",
			createdAt: "2024-01-02",
		},
	];

	beforeEach(() => {
		jest.clearAllMocks();

		const photosStore = mockPhotosStore.getState();
		photosStore.photos = mockPhotos;
		photosStore.isPhotosLoading = false;
		photosStore.hasMore = true;
		photosStore.isAddPhotosLoading = false;
		photosStore.reset = jest.fn();
		photosStore.getAllPhotos = jest.fn().mockResolvedValue(mockPhotos);

		const userStore = mockUserStore.getState();
		userStore.user = {
			id: "123",
			firstName: "John",
			secondName: "",
			thirdName: "Doe",
			email: "john@example.com",
			phone: "+1234567890",
			avatarUrl: "",
			avatarCreateDate: "",
			fullName: "John Doe",
		};

		mockPhotosService.photos = mockPhotos;
		mockPhotosService.count = mockPhotos.length;
		mockPhotosService.getAllPhotos = jest.fn().mockResolvedValue(mockPhotos);
		mockPhotosService.onClickPhoto = jest.fn();
		mockPhotosService.deletePhoto = jest.fn();
		mockPhotosService.addPhotos = jest.fn();
		mockPhotosService.addPhotoFromAvatar = jest.fn();

		const profileStore = mockProfileStore.getState();
		profileStore.isPrepareAnotherUser = false;
		profileStore.isMe = true;
	});

	it("matches snapshot in initial state", async () => {
		const { container } = await act(async () => {
			return render(<Photos />);
		});
		expect(container).toMatchSnapshot();
	});

	it("matches snapshot with loading state", async () => {
		const photosStore = mockPhotosStore.getState();
		photosStore.isPhotosLoading = true;

		const { container } = await act(async () => {
			return render(<Photos />);
		});
		expect(container).toMatchSnapshot();
	});

	it("matches snapshot with empty state", async () => {
		const photosStore = mockPhotosStore.getState();
		photosStore.photos = [];
		photosStore.getAllPhotos = jest.fn().mockResolvedValue([]);

		const { container } = await act(async () => {
			return render(<Photos />);
		});
		expect(container).toMatchSnapshot();
	});

	it("renders photos grid correctly", async () => {
		const { container } = await act(async () => {
			return render(<Photos />);
		});

		const photosContainer = container.querySelector(".photos__container");
		expect(photosContainer).toBeInTheDocument();

		const photoElements = container.querySelectorAll(".photos__container__item");
		expect(photoElements).toHaveLength(mockPhotos.length);

		mockPhotos.forEach((photo, index) => {
			const imgElement = photoElements[index].querySelector("img");
			expect(imgElement).toHaveAttribute("src", expect.stringContaining(photo.path));
		});
	});

	it("shows loading state", async () => {
		const photosStore = mockPhotosStore.getState();
		photosStore.isPhotosLoading = true;

		const { container } = await act(async () => {
			return render(<Photos />);
		});

		const spinner = container.querySelector(".spinner");
		expect(spinner).toBeInTheDocument();
	});

	it("shows empty state when no photos", async () => {
		const photosStore = mockPhotosStore.getState();
		photosStore.photos = [];
		photosStore.getAllPhotos = jest.fn().mockResolvedValue([]);

		const { container } = await act(async () => {
			return render(<Photos />);
		});

		const emptyState = container.querySelector(".no-data");
		expect(emptyState).not.toBeInTheDocument();
	});

	it("handles photo click", async () => {
		const { container } = await act(async () => {
			return render(<Photos />);
		});

		const firstPhoto = container.querySelector(".photo__image");
		expect(firstPhoto).toBeInTheDocument();
		fireEvent.click(firstPhoto!);

		expect(mockPhotosService.onClickPhoto).toHaveBeenCalledWith(mockPhotos[0].id);
	});

	it("handles photo deletion", async () => {
		const { container } = await act(async () => {
			return render(<Photos />);
		});

		const photo = container.querySelector(".photo__image");
		expect(photo).toBeInTheDocument();
		fireEvent.click(photo!);

		expect(mockPhotosService.deletePhoto).not.toHaveBeenCalled();
	});
});
