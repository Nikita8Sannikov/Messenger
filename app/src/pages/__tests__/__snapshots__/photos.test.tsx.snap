// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Photos Component matches snapshot in initial state 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 photos paper-block css-1gtchvp-MuiPaper-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
  >
    <div
      class="photos__container MuiBox-root css-0"
      data-testid="box-component"
    >
      <div
        style="position: relative; height: 0px; width: 0px; overflow: auto; will-change: transform; direction: ltr;"
      >
        <div
          style="height: 516px; width: 100%;"
        >
          <div
            class="list__row"
            style="position: absolute; left: 0px; top: 0px; height: 258px; width: 100%; gap: 8px;"
          >
            <div>
              <div
                class="photos__container__item"
                style="width: -5.333333333333333px; height: 250px;"
              >
                <div
                  class="photo"
                >
                  <img
                    alt="1"
                    class="photo__image"
                    data-testid="image"
                    loading="lazy"
                    src="/path/to/photo1.jpg?w=-10.666666666666666"
                  />
                </div>
              </div>
            </div>
            <div>
              <div
                class="photos__container__item"
                style="width: -5.333333333333333px; height: 250px;"
              >
                <div
                  class="photo"
                >
                  <img
                    alt="2"
                    class="photo__image"
                    data-testid="image"
                    loading="lazy"
                    src="/path/to/photo2.jpg?w=-10.666666666666666"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="list__loading-row"
            style="position: absolute; left: 0px; top: 258px; height: 258px; width: 100%;"
          >
            <div
              class="spinner"
              data-testid="spinner"
            >
              <span
                class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1wxecgq-MuiCircularProgress-root"
                role="progressbar"
                style="width: 40px; height: 40px;"
              >
                <svg
                  class="MuiCircularProgress-svg css-54pwck-MuiCircularProgress-svg"
                  viewBox="22 22 44 44"
                >
                  <circle
                    class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-19t5dcl-MuiCircularProgress-circle"
                    cx="44"
                    cy="44"
                    fill="none"
                    r="20.2"
                    stroke-width="3.6"
                  />
                </svg>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Photos Component matches snapshot with empty state 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 photos paper-block css-1gtchvp-MuiPaper-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
  >
    <div
      class="photos__container MuiBox-root css-0"
      data-testid="box-component"
    >
      <div
        class="virtual-list__no-content opacity-text"
      >
        There are no photos yet.
      </div>
    </div>
  </div>
</div>
`;

exports[`Photos Component matches snapshot with loading state 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 photos paper-block css-1gtchvp-MuiPaper-root"
    style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
  >
    <div
      class="photos__container MuiBox-root css-0"
      data-testid="box-component"
    >
      <div
        style="position: relative; height: 0px; width: 0px; overflow: auto; will-change: transform; direction: ltr;"
      >
        <div
          style="height: 516px; width: 100%;"
        >
          <div
            class="list__row"
            style="position: absolute; left: 0px; top: 0px; height: 258px; width: 100%; gap: 8px;"
          >
            <div>
              <div
                class="photos__container__item"
                style="width: -5.333333333333333px; height: 250px;"
              >
                <div
                  class="photo"
                >
                  <img
                    alt="1"
                    class="photo__image"
                    data-testid="image"
                    loading="lazy"
                    src="/path/to/photo1.jpg?w=-10.666666666666666"
                  />
                </div>
              </div>
            </div>
            <div>
              <div
                class="photos__container__item"
                style="width: -5.333333333333333px; height: 250px;"
              >
                <div
                  class="photo"
                >
                  <img
                    alt="2"
                    class="photo__image"
                    data-testid="image"
                    loading="lazy"
                    src="/path/to/photo2.jpg?w=-10.666666666666666"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="list__loading-row"
            style="position: absolute; left: 0px; top: 258px; height: 258px; width: 100%;"
          >
            <div
              class="spinner"
              data-testid="spinner"
            >
              <span
                class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-1wxecgq-MuiCircularProgress-root"
                role="progressbar"
                style="width: 40px; height: 40px;"
              >
                <svg
                  class="MuiCircularProgress-svg css-54pwck-MuiCircularProgress-svg"
                  viewBox="22 22 44 44"
                >
                  <circle
                    class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-19t5dcl-MuiCircularProgress-circle"
                    cx="44"
                    cy="44"
                    fill="none"
                    r="20.2"
                    stroke-width="3.6"
                  />
                </svg>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
